## ebay Used Car Sales
In this guided project, we'll work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.

The dataset was originally scraped and uploaded to Kaggle by user orgesleka.
The original dataset isn't available on Kaggle anymore.
The data dictionary provided with data is as follows:

* dateCrawled - When this ad was first crawled. All field-values are taken from this date.
* name - Name of the car.
* seller - Whether the seller is private or a dealer.
* offerType - The type of listing
* price - The price on the ad to sell the car.
* abtest - Whether the listing is included in an A/B test.
* vehicleType - The vehicle Type.
* yearOfRegistration - The year in which the car was first registered.
* gearbox - The transmission type.
* powerPS - The power of the car in PS.
* model - The car model name.
* kilometer - How many kilometers the car has driven.
* monthOfRegistration - The month in which the car was first registered.
* fuelType - What type of fuel the car uses.
* brand - The brand of the car.
* notRepairedDamage - If the car has a damage which is not yet repaired.
* dateCreated - The date on which the eBay listing was created.
* nrOfPictures - The number of pictures in the ad.
* postalCode - The postal code for the location of the vehicle.
* lastSeenOnline - When the crawler saw this ad last online.

The aim of this project is to clean the data and analyze the included used car listings. 

In [1]:
import pandas as pd
import numpy as np
autos = pd.read_csv('autos.csv', encoding = 'Latin-1')

In [2]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [3]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [4]:
autos.columns = 'date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest', 'vehicle_type', 'registration_year', 'gearbox', 'power _ps', 'model', 'odometer', 'registration_month', 'fuel_type', 'brand', 'unepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code', 'last_seen'
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power _ps   model   odometer  registration_month fuel_type  \
0    manuell        158  andere  150,000km                   3       lpg   
1  automatik        286     7er  150,000km                   6    benzin   
2    manuell        102    golf   70,000km                   7    benzin   
3  automatik         71  fortwo   70,000km                   6    benzin   
4    manuell          0   focus  150,000km                   7    benzin   

        brand unepaired_damage           ad_created  nr_of_pictures  \
0     peugeot             nein  2016-03-26 00:00:00               0   
1         bmw             nein  2016-04-04 00:00:00               0   
2  volkswagen             nein  2016-03-26 00:00:00               0   
3       smart             nein  2016-03-12 00:00:00               0   
4        ford             nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

We went ahead and changed the colun names from camelcase to Python's preferred snakcase and changed some of the names to be more descriptive of their contents.

In [5]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-23 18:39:34  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox     power _ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unepaired_damage  \
count       50000        50000.000000     45518       50000            40171   
unique         13                 NaN         7          40                2   
top     150,000km                 NaN    benzin  volkswagen             nein   
freq        32424                 NaN     30107       10687            35232   
mean          NaN            5.723360       NaN         NaN              NaN   
std           NaN            3.711984       NaN         NaN              NaN   
min           NaN            0.000000       NaN         NaN              NaN   
25%           NaN            3.000000       NaN         NaN              NaN   
50%           NaN            6.000000       NaN         NaN              NaN   
75%           NaN            9.000000       NaN         NaN              NaN   
max           NaN           12.000000       NaN         NaN              NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

It seems as though the nr_of_pictures column is mostly the same number, this isn't useful to our analytics so we can drop it. We can also see that the max registration year is 9999, that is something to investigate further. Also both price and odometer are not included here because they have string types so we can clean that as well.

In [6]:
autos['odometer'] = autos['odometer'].str.replace('km','').str.replace(',', '').astype(int)
autos['price'] = autos['price'].str.replace('$','').str.replace(',', '').astype(float)

In [7]:
autos.rename({'odometer': 'odometer_km'},axis=1,inplace=True)
autos.rename({'price': 'price_dollars'},axis=1,inplace=True)

In [8]:
autos

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offer_type  price_dollars   abtest vehicle_type  \
0      privat    Angebot         5000.0  control          bus   
1      privat    Angebot         8500.0  control    limousine   
2      privat    Angebot         8990.0     test    limousine   
3      privat    Angebot         4350.0  control   kleinwagen   
4      privat    Angebot         1350.0     test        kombi   
...       ...        ...            ...      ...          ...   
49995  privat    Angebot        24900.0  control    limousine   
49996  privat    Angebot         1980.0  control       cabrio   
49997  privat    Angebot        13200.0     test       cabrio   
49998  privat    Angebot        22900.0  control        kombi   
49999  privat    Angebot         1250.0  control    limousine   

       registration_year    gearbox  power _ps   model  odometer_km  \
0                   2004    manuell        158  andere       150000   
1                   1997  automatik        286     7er       150000   
2                   2009    manuell        102    golf        70000   
3                   2007  automatik         71  fortwo        70000   
4                   2003    manuell          0   focus       150000   
...                  ...        ...        ...     ...          ...   
49995               2011  automatik        239      q5       100000   
49996               1996    manuell         75   astra       150000   
49997               2014  automatik         69     500         5000   
49998               2013    manuell        150      a3        40000   
49999               1996    manuell        101  vectra       150000   

       registration_month fuel_type       brand unepaired_damage  \
0                       3       lpg     peugeot             nein   
1                       6    benzin         bmw             nein   
2                       7    benzin  volkswagen             nein   
3                       6    benzin       smart             nein   
4                       7    benzin        ford             nein   
...                   ...       ...         ...              ...   
49995                   1    diesel        audi             nein   
49996                   5    benzin        opel             nein   
49997                  11    benzin        fiat             nein   
49998                  11    diesel        audi             nein   
49999                   1    benzin        opel             nein   

                ad_created  nr_of_pictures  postal_code            last_seen  
0      2016-03-26 00:00:00               0        79588  2016-04-06 06:45:54  
1      2016-04-04 00:00:00               0        71034  2016-04-06 14:45:08  
2      2016-03-26 00:00:00               0        35394  2016-04-06 20:15:37  
3      2016-03-12 00:00:00               0        33729  2016-03-15 03:16:28  
4      2016-04-01 00:00:00               0        39218  2016-04-01 14:38:50  
...                    ...             ...          ...                  ...  
49995  2016-03-27 00:00:00           

In [9]:
print(autos['odometer_km'].unique().shape)
print('\n')
print(autos['odometer_km'].describe())
print('\n')
print(autos['odometer_km'].value_counts().sort_index(ascending=False))
print('\n')
print(autos['price_dollars'].unique().shape)
print('\n')
print(autos['price_dollars'].describe())
print('\n')
print(autos['price_dollars'].value_counts().sort_index(ascending=False))

(13,)


count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
20000       784
10000       264
5000        967
Name: odometer_km, dtype: int64


(2357,)


count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_dollars, dtype: float64


99999999.0       1
27322222.0       1
12345678.0       3
11111111.0       2
10000000.0       1
              ... 
5.0              2
3.0              1
2.0              3
1.0            156
0.0           1421
Name: price_dollars, Length: 2357, dtype: int64


In [10]:
autos['price_dollars'] = autos.loc[autos['price_dollars'].between(500, 350000), 'price_dollars']
autos['odometer_km'] = autos.loc[autos['odometer_km'].between(10000, 150000), 'odometer_km']

The price increases steadily until 350000 then it increaes rapidly so we can use 350000 as an upper limit,The lowest price is zero, which has 1421 occurences in the dataset. 500 seems like a reasonable lower limit for this column.

In [11]:
print(autos['odometer_km'].unique().shape)
print('\n')
print(autos['odometer_km'].describe())
print('\n')
print(autos['odometer_km'].value_counts().sort_index(ascending=False))
print('\n')
print(autos['price_dollars'].unique().shape)
print('\n')
print(autos['price_dollars'].describe())
print('\n')
print(autos['price_dollars'].value_counts().sort_index(ascending=False))

(13,)


count     49033.000000
mean     128113.719332
std       36631.381785
min       10000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


150000.0    32424
125000.0     5170
100000.0     2169
90000.0      1757
80000.0      1436
70000.0      1230
60000.0      1164
50000.0      1027
40000.0       819
30000.0       789
20000.0       784
10000.0       264
Name: odometer_km, dtype: int64


(2209,)


count     45097.000000
mean       6320.659600
std        9261.841444
min         500.000000
25%        1500.000000
50%        3500.000000
75%        7900.000000
max      350000.000000
Name: price_dollars, dtype: float64


350000.0      1
345000.0      1
299000.0      1
295000.0      1
265000.0      1
           ... 
520.0         8
517.0         1
510.0         2
501.0         1
500.0       781
Name: price_dollars, Length: 2208, dtype: int64


In [12]:
autos

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offer_type  price_dollars   abtest vehicle_type  \
0      privat    Angebot         5000.0  control          bus   
1      privat    Angebot         8500.0  control    limousine   
2      privat    Angebot         8990.0     test    limousine   
3      privat    Angebot         4350.0  control   kleinwagen   
4      privat    Angebot         1350.0     test        kombi   
...       ...        ...            ...      ...          ...   
49995  privat    Angebot        24900.0  control    limousine   
49996  privat    Angebot         1980.0  control       cabrio   
49997  privat    Angebot        13200.0     test       cabrio   
49998  privat    Angebot        22900.0  control        kombi   
49999  privat    Angebot         1250.0  control    limousine   

       registration_year    gearbox  power _ps   model  odometer_km  \
0                   2004    manuell        158  andere     150000.0   
1                   1997  automatik        286     7er     150000.0   
2                   2009    manuell        102    golf      70000.0   
3                   2007  automatik         71  fortwo      70000.0   
4                   2003    manuell          0   focus     150000.0   
...                  ...        ...        ...     ...          ...   
49995               2011  automatik        239      q5     100000.0   
49996               1996    manuell         75   astra     150000.0   
49997               2014  automatik         69     500          NaN   
49998               2013    manuell        150      a3      40000.0   
49999               1996    manuell        101  vectra     150000.0   

       registration_month fuel_type       brand unepaired_damage  \
0                       3       lpg     peugeot             nein   
1                       6    benzin         bmw             nein   
2                       7    benzin  volkswagen             nein   
3                       6    benzin       smart             nein   
4                       7    benzin        ford             nein   
...                   ...       ...         ...              ...   
49995                   1    diesel        audi             nein   
49996                   5    benzin        opel             nein   
49997                  11    benzin        fiat             nein   
49998                  11    diesel        audi             nein   
49999                   1    benzin        opel             nein   

                ad_created  nr_of_pictures  postal_code            last_seen  
0      2016-03-26 00:00:00               0        79588  2016-04-06 06:45:54  
1      2016-04-04 00:00:00               0        71034  2016-04-06 14:45:08  
2      2016-03-26 00:00:00               0        35394  2016-04-06 20:15:37  
3      2016-03-12 00:00:00               0        33729  2016-03-15 03:16:28  
4      2016-04-01 00:00:00               0        39218  2016-04-01 14:38:50  
...                    ...             ...          ...                  ...  
49995  2016-03-27 00:00:00           

Right now, the date_crawled, last_seen, and ad_created columns are all identified as string values by pandas. Because these three columns are represented as strings, we need to convert the data into a numerical representation so we can understand it quantitatively. The other two columns are represented as numeric values, so we can use methods like Series.describe() to understand the distribution without any extra data processing.

In [13]:
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).mul(100).round(1).sort_index()

2016-03-05    2.5
2016-03-06    1.4
2016-03-07    3.6
2016-03-08    3.3
2016-03-09    3.3
2016-03-10    3.2
2016-03-11    3.2
2016-03-12    3.7
2016-03-13    1.6
2016-03-14    3.7
2016-03-15    3.4
2016-03-16    2.9
2016-03-17    3.2
2016-03-18    1.3
2016-03-19    3.5
2016-03-20    3.8
2016-03-21    3.8
2016-03-22    3.3
2016-03-23    3.2
2016-03-24    2.9
2016-03-25    3.2
2016-03-26    3.2
2016-03-27    3.1
2016-03-28    3.5
2016-03-29    3.4
2016-03-30    3.4
2016-03-31    3.2
2016-04-01    3.4
2016-04-02    3.5
2016-04-03    3.9
2016-04-04    3.7
2016-04-05    1.3
2016-04-06    0.3
2016-04-07    0.1
Name: date_crawled, dtype: float64

It seems as though the crawler was active for about a month from March-April 2016

In [14]:
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).mul(100).round(1).sort_index()

2015-06-11    0.0
2015-08-10    0.0
2015-09-09    0.0
2015-11-10    0.0
2015-12-05    0.0
             ... 
2016-04-03    3.9
2016-04-04    3.7
2016-04-05    1.2
2016-04-06    0.3
2016-04-07    0.1
Name: ad_created, Length: 76, dtype: float64

The dates most of the adds were created occur after the beginning of the crawl which makes sense, ebay sales are usually only active for short amounts of time.

In [15]:
autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).mul(100).round(1).sort_index()

2016-03-05     0.1
2016-03-06     0.4
2016-03-07     0.5
2016-03-08     0.8
2016-03-09     1.0
2016-03-10     1.1
2016-03-11     1.3
2016-03-12     2.4
2016-03-13     0.9
2016-03-14     1.3
2016-03-15     1.6
2016-03-16     1.6
2016-03-17     2.8
2016-03-18     0.7
2016-03-19     1.6
2016-03-20     2.1
2016-03-21     2.1
2016-03-22     2.2
2016-03-23     1.9
2016-03-24     2.0
2016-03-25     1.9
2016-03-26     1.7
2016-03-27     1.6
2016-03-28     2.1
2016-03-29     2.2
2016-03-30     2.5
2016-03-31     2.4
2016-04-01     2.3
2016-04-02     2.5
2016-04-03     2.5
2016-04-04     2.5
2016-04-05    12.4
2016-04-06    22.1
2016-04-07    13.1
Name: last_seen, dtype: float64

Most of the last_seen dates occur at the end of the distribution which makes sense as the dates are located right at the end of the crawl.

In [16]:
autos['registration_year'].value_counts(normalize=True, dropna=False).mul(100).round(1).sort_index()

1000    0.0
1001    0.0
1111    0.0
1500    0.0
1800    0.0
       ... 
6200    0.0
8888    0.0
9000    0.0
9996    0.0
9999    0.0
Name: registration_year, Length: 97, dtype: float64

One thing that stands out from the exploration we did in the last screen is that the registration_year column contains some odd values:

* The minimum value is 1000, before cars were invented
* The maximum value is 9999, many years into the future

Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. 

Let's count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely, or if we need more custom logic.

In [17]:
autos['registration_year'] = autos.loc[autos['registration_year'].between(1900, 2016), 'registration_year']

In [18]:
autos['registration_year'].describe()

count    48028.00000
mean      2002.80351
std          7.31085
min       1910.00000
25%       1999.00000
50%       2003.00000
75%       2008.00000
max       2016.00000
Name: registration_year, dtype: float64

Seems as though we have a mean distribution of 2002

In [19]:
autos['brand'].value_counts()

volkswagen        10687
opel               5461
bmw                5429
mercedes_benz      4734
audi               4283
ford               3479
renault            2404
peugeot            1456
fiat               1308
seat                941
skoda               786
mazda               757
nissan              754
citroen             701
smart               701
toyota              617
sonstige_autos      546
hyundai             488
volvo               457
mini                424
mitsubishi          406
honda               399
kia                 356
alfa_romeo          329
porsche             294
suzuki              293
chevrolet           283
chrysler            181
dacia               129
daihatsu            128
jeep                110
subaru              109
land_rover           99
saab                 80
daewoo               79
trabant              78
jaguar               77
rover                69
lancia               57
lada                 31
Name: brand, dtype: int64

In [20]:
top_20 = autos['brand'].value_counts().head(20).index
top_20

Index(['volkswagen', 'opel', 'bmw', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'mazda', 'nissan', 'citroen',
       'smart', 'toyota', 'sonstige_autos', 'hyundai', 'volvo', 'mini'],
      dtype='object')

In [21]:
price_brand = {}
for brand in top_20:
    mean_price = round(autos.loc[autos['brand'] == brand, 'price_dollars'].mean())
    price_brand[brand] = mean_price
print(price_brand)

{'volkswagen': 5701.0, 'opel': 3349.0, 'bmw': 8515.0, 'mercedes_benz': 8670.0, 'audi': 9484.0, 'ford': 4267.0, 'renault': 2761.0, 'peugeot': 3329.0, 'fiat': 3212.0, 'seat': 4710.0, 'skoda': 6540.0, 'mazda': 4392.0, 'nissan': 5077.0, 'citroen': 3995.0, 'smart': 3552.0, 'toyota': 5244.0, 'sonstige_autos': 13201.0, 'hyundai': 5695.0, 'volvo': 5111.0, 'mini': 10617.0}


In [22]:
odo = {}
for brand in top_20:
    mean_odo = round(autos.loc[autos['brand'] == brand, 'odometer_km'].mean())
    odo[brand] = mean_odo

In [23]:
bmp_series = pd.Series(price_brand)
bmo_series = pd.Series(odo)

In [25]:
brand_means = pd.DataFrame(bmp_series, columns=['mean_price'])
brand_means['mean_mileage'] = bmo_series
print(brand_means)

                mean_price  mean_mileage
volkswagen          5701.0      131187.0
opel                3349.0      132091.0
bmw                 8515.0      134817.0
mercedes_benz       8670.0      132721.0
audi                9484.0      130967.0
ford                4267.0      126436.0
renault             2761.0      130734.0
peugeot             3329.0      128540.0
fiat                3212.0      120208.0
seat                4710.0      123958.0
skoda               6540.0      113009.0
mazda               4392.0      127067.0
nissan              5077.0      121135.0
citroen             3995.0      121257.0
smart               3552.0      101444.0
toyota              5244.0      117262.0
sonstige_autos     13201.0      101298.0
hyundai             5695.0      107836.0
volvo               5111.0      139812.0
mini               10617.0       91205.0


Minis tend to be the most expensive but also tend toward having the least mileage, just comparing these two columns it seems as though Hyunda